File: LeetCode 448 - Day 1/day1.cpp
// 448. Find All Numbers Disappeared in an Array

#include <iostream>
#include <vector>
using namespace std;

// class declaration
class Solution
{

    // public declaration - this is for function and variables
public:
    // function declaration - returns a vector<int> and takes in a vector<int>
    // vectors are dynamic arrays with the ability to resize itself when an object is inserted and deleted
    // & sign means memory address - need to look into this
    vector<int> findDisappearedNumbers(vector<int> &nums)
    {
        // int len - size of arg nums
        int len = nums.size(); // 8
        // loop through len
        for (int i = 0; i < len; i++)
        {
            int m = abs(nums[i]) - 1;
            nums[m] = nums[m] > 0 ? -nums[m] : nums[m];
        }
        // create a vector res
        vector<int> res;
        // loop through len
        for (int i = 0; i < len; i++)
        {
            // if x is greater than 0 add i + 1 to res
            if (nums[i] > 0)
                res.push_back(i + 1);
        }
        return res;
    }
};

// -4 3 2 7 -8 2 -3 -1
// 2 3 4 6

/*
LeetCode Easy
Loop through len of vector and negate something 
Loop through again and look for something - if not there add to result
*/

File: LeetCode 448 - Day 1/day1.cs
// 448. Find All Numbers Disappeared in an Array

using System;

namespace Day1
{
    class Solution {

        // function name 
        public IList<int> findDisappearedNumbers(int[] nums) {
            // init vars
            HashSet<int> list = new HashSet<int>();
            List<int> result = new List<int>();

            // loop through nums length and add to list
            for(int i = 0; i < nums.Length; i++) {
                list.Add(nums[i]);
            }

            // if 1 - nums.length is not in list add to result 
            for(int i = 1; i < nums.Length; i++) {
                if(!list.Contains(i)) {
                    result.Add(i);
                }
            }
            return result;
        }
    }
    
}

/*
This structure is more like java and makes more sense to me right now.
Also im not sure why the methods are capitalized, need to look into this
*/

